peermanager/peermanager.go

```diff
@@ -189,22 +99,56 @@ func (pm *PeerManager) stopPeerHandler(p peer.ID) {
 	}
 
 	pq.Shutdown()
-	delete(pm.peerQueues, p)
+
+	pm.remove(p)
 }
 
-func (pm *PeerManager) sendMessage(ms *sendPeerMessage) {
-	if len(ms.targets) == 0 {
-		for _, p := range pm.peerQueues {
-			p.AddMessage(ms.entries, ms.from)
-		}
+// SendMessage is called to send a message to all or some peers in the pool;
+// if targets is nil, it sends to all.
+func (pm *PeerManager) SendMessage(entries []*bsmsg.Entry, targets []peer.ID, from uint64) {
+	if len(targets) == 0 {
+		pm.iterate(func(_ peer.ID, p PeerQueue) {
+			p.AddMessage(entries, from)
+		})
 	} else {
-		for _, t := range ms.targets {
-			p, ok := pm.peerQueues[t]
+		for _, t := range targets {
+			p, ok := pm.get(t)
 			if !ok {
-				log.Infof("tried sending wantlist change to non-partner peer: %s", t)
-				continue
+				p = pm.createPeerQueue(t)
+				pm.set(t, p)
```

Comment:
Can nothing call this concurrently? If so, I assume we'd need some kind of atomic GetOrCreate kind of thing.